crates/oxc_parser/src/lexer/source.rs

```diff
@@ -0,0 +1,405 @@
+        // so slice from `ptr` to `end` will always be a valid UTF-8 string.
+        unsafe {
+            let len = self.end as usize - self.ptr as usize;
+            let slice = slice::from_raw_parts(self.ptr, len);
+            debug_assert!(slice.is_empty() || !is_utf8_cont_byte(slice[0]));
+            str::from_utf8_unchecked(slice)
+        }
+    }
+
+    // Return whether at end of source.
+    #[inline]
+    pub(super) fn is_eof(&self) -> bool {
+        self.ptr == self.end
+    }
+
+    /// Get current position.
+    ///
+    /// The `SourcePosition` returned is guaranteed to be within bounds of `&str` that `Source`
+    /// was created from, and on a UTF-8 character boundary, so can be used by caller
+    /// to later move current position of this `Source` using `Source::set_position`.
+    ///
+    /// `SourcePosition` lives as long as the source text `&str` that `Source` was created from.
+    #[inline]
+    pub(super) fn position(&self) -> SourcePosition<'a> {
+        SourcePosition { ptr: self.ptr, _marker: PhantomData }
+    }
+
+    /// Move current position.
+    ///
+    /// # SAFETY
+    /// `pos` must be created from this `Source`, not another `Source`.
+    /// If this is the case, the invariants of `Source` are guaranteed to be upheld.
+    #[inline]
+    pub(super) unsafe fn set_position(&mut self, pos: SourcePosition) {
+        // `SourcePosition` always upholds the invariants of `Source`,
+        // as long as it's created from this `Source`.
+        // SAFETY: `read_u8`'s contract is upheld by:
+        // * The preceding checks that `pos.ptr` >= `self.start` and < `self.end`.
+        // * `Source`'s invariants guarantee that `self.start` - `self.end` contains allocated memory.
+        // * `Source::new` takes an immutable ref `&str`, guaranteeing that the memory `pos.ptr`
+        //   addresses cannot be aliased by a `&mut` ref as long as `Source` exists.
+        // * `SourcePosition` can only live as long as the `&str` underlying `Source`.
+        debug_assert!(
+            pos.ptr >= self.start
+                && pos.ptr <= self.end
+                && (pos.ptr == self.end || !is_utf8_cont_byte(read_u8(pos.ptr)))
+        );
+        self.ptr = pos.ptr;
+    }
+
+    /// Get current position in source, relative to start of source.
+    #[allow(clippy::cast_possible_truncation)]
+    #[inline]
+    pub(super) fn offset(&self) -> u32 {
+        // Cannot overflow `u32` because of `MAX_LEN` check in `Source::new`
+        (self.ptr as usize - self.start as usize) as u32
+    }
+
+    /// Move current position back by `n` bytes.
+    ///
+    /// # Panic
+    /// Panics if:
+    /// * `n` is 0.
+    /// * `n` is greater than current offset in source.
+    /// * Moving back `n` bytes would not place current position on a UTF-8 character boundary.
+    #[inline]
+    pub(super) fn back(&mut self, n: usize) {
+        assert!(n > 0, "Cannot call `Source::back` with 0");
```

Comment:
Possibly best to comment these `assert`s are intentional for compiler hints.

Reply:
That assert is actually essential to make the function safe, rather than a compiler hint. But yes you're right, it wasn't clear why it was there, and needed a comment. Have added one.

Reply:
> to make the function safe

All the other functions use `debug_assert`, so I couldn't tell whether all the `asserts` need to be here in the release build.